### ISSUE TITLE
Feature request: Validate modified fields

### ISSUE THREAD

**Issue by vfonic** [vazco/uniforms#510]
I searched the issues and couldn't find similar one.

My client wants to have the validation fields appear immediately as user is typing, but only for those fields that have been modified (dirty/changed).

Current validation modes don't support this. `onChange` validates *all* fields immediately. `onChangeAfterSubmit` doesn't start validating immediately, `onSubmit` only validates on every submit event.

I'd like a fourth option `onChangePerField` or `onFieldChange` or `onChangeForDirtyFields` or `onChangeForModifiedFields` or `onChangeForChangedFields`.

I've already seen this type of validation in other web apps. I think it might be added to the main package.

We could use `changedMap` from uniforms context to get all the fields that need to be validated. I didn't find a simple way to access the uniforms context from `onValidate` method.

What do you think?

**Comment by radekmie** [comment]
Hi @vfonic. That's a tricky one! Okay, let's get to the details:

* Most schemas cannot validate a part of a doc, so it still would be a full validation.
* Extracting field scoped error from the validation error happens on the schema level.
* Props defining the validation scheme and `changedMap` are on the form level.
* Errors are shown in the fields, most of them are using the `BaseField` logic and a `showInlineError` flag.

Combining the above leaves only two options: either somehow incorporate this within fields to know about it or somehow force the form/schema to silence unwanted errors.

There is also one more tough question: where to put it to allow every field work independently (i.e. one field shows an error all the time and the other only after a change).

Ideas?

**Comment by vfonic** [comment]
> * Most schemas cannot validate a part of a doc, so it still would be a full validation.

That's true, but that is related to specific schema features. I think we could offer this as an option and mention that it doesn't work with those schemas, or how to make it work with those schemas.

> * Extracting field scoped error from the validation error happens on the schema level.

It could also happen outside of schema. You can just do something like:
`errors['description'] = 'Description should be at least 100 characters'`
...I guess?

> * Props defining the validation scheme and `changedMap` are on the form level.

Is it possible to get uniforms context in `onValidate` method? We could then use that method to get the errors for `changedMap` from the context.

> * Errors are shown in the fields, most of them are using the `BaseField` logic and a `showInlineError` flag.

Not sure if this would work with the above-mentioned way to set errors:
`errors['description'] = 'Description should be at least 100 characters'`
?

**Comment by radekmie** [comment]
> That's true, but that is related to specific schema features. I think we could offer this as an option and mention that it doesn't work with those schemas, or how to make it work with those schemas.

I only know a single schema that can do it (ajv) and it still validates up to the first error, not a given field.

> It could also happen outside of schema. [...]

No, I meant the `Bridge.getError(name, error) -> ?error`. The bridge is responsible for getting an error for a given field, called `name`, out of a validation error `error`.

> Is it possible to get uniforms context in `onValidate` method? We could then use that method to get the errors for `changedMap` from the context.

Not directly. You can use `ref` to get the form `state`, but it's not a clean solution - rather a workaround.

> Not sure if this would work with the above-mentioned way to set errors: [...]

No, I meant [this code in `BaseField.getFieldProps`](https://github.com/vazco/uniforms/blob/7e96b812c95ef3bfd561c432b3baec7debc02a42/packages/uniforms/src/BaseField.js#L195). Fields are not using the bridge nor the schema directly.

**Comment by vfonic** [comment]
I was looking into passing the context over here:

https://github.com/vazco/uniforms/blob/547bb3bc2ccfbbe6700a85c91e5c724c7d2c458f/packages/uniforms/src/AutoForm.js#L75-L77

But I can't get the local version of the package to work in my project. :/

I've opened another issue for that https://github.com/vazco/uniforms/issues/511
Hope that makes sense as it's a separate issue.

**Comment by radekmie** [comment]
You don't have to do it - simply extend the `AutoForm` in your project. I'll leave it open for now, as the problem is still there, and it has to be _at least_ answered.

**Comment by radekmie** [comment]
I still have no idea, on how to tackle it, but it shouldn't be a breaking change (at least it doesn't look like one). The problem is still open and waiting for suggestions.

**Comment by vfonic** [comment]
Some requirements changed from my client and I don't need this
functionality right now. If I need it again, I could look into it. Thanks!

**Comment by radekmie** [comment]
I'm cleaning the issues up a little bit and here's a resolution: I don't see any _clean_ solution, other than adding an additional parameter here and there, which can be done in your project, without any core change. I'm closing, but if something comes up - feel free to write it down here.
